### ISSUE TITLE
Using Orbiter 2 sensor with Happy Hare (detection delay issue)

### ISSUE THREAD

**Issue by tiaanv** [RobertLorincz/Orbiter-2-Smart-Sensor#1]
I have fitted the Orbiter 2 smart sensor to my tool head and have recently added an MMU into the fold.  I am trying to use the O2 sensor as my extruder sensor (sensor before extruder), but I have hit a bit of a snag.

The sensor has a built-in delay from when it triggers internally to when it reports this to the sensor pin.  Not sure what the exact delay is, but it's quite long.  I am sure this is by design and makes perfect sense for most standalone applications,  However, in the case of using it as a sensor (part of of complete system), This delay is problematic.

For example, when calibrating my tool head sensor and nozzle distances, this delay causes an error when homing to the extruder sensor, because the sensor triggers much too late..

My questions:
1.  Can this delay be disabled in any way?
2.  Have you come across this use case before, and has someone come up with a solution?

**Comment by RobertLorincz** [comment]
The hardware filter time/delay is one second. I have implemented a filter time over the filament sense signal of one second. The main reason is to avoid double triggering of the sensor by blobs or any glitches. In the past, many users had issues with fake runout detection caused by minor glotches. The aim of this HW filter is to reduce this effect. By default, the klipper filament runnout swicth section has a detection delay configured of one second as well. You may set that to zero, effectively halfening the delay you are experiencing.

**Comment by RobertLorincz** [comment]
I uploaded a corrected configuration with event delay set to 0.1 plus another bug correction. Let me know if this improves your setup. In the end, having HW filter, we do not necessarily need any additional software filters.

**Comment by tiaanv** [comment]
Thx for the response Robert.  

I don't have this configured as a runout sensor, so there is no delay configured. It is registered as the "extruder pin" extruder_switch_pin: EBBCan: PB3  in happy hare.  The built-in delay of the board interferes with HH's running of the homing logic.  But I assume it would not be easy to re-program the board and disable your built-in filtering, or at least make it only a debounce filter with a delay of less than 100ms.  For now, I have taken the sensor out of my configuration completely.  I might end up removing the sensor altogether since it no longer adds value to my specific intended use case.  Just wanted to check in if there was a way to disable the filtering/delay.

**Comment by tiaanv** [comment]
FWIW, I think the filtering is a good idea, but would argue that this type of delay can inherently be handled by klipper, and baking it into the board firmware puts it outside of the control of the user.  perhaps for future firmware revision, a simple power-up while holding down the unload_button, to disable the filter logic (and save to eeprom) could be feasible?

**Comment by RobertLorincz** [comment]
That would be feasible; the micro does have an integrated EEPROM to save such settings. My concern is that we will shoot back to the old glitch and stuff issues we had in the past. At least some issue klipper was not able to cope with all such cases. This was the main reason I implemented this signal conditioning in the sensor. Sure, maybe 1 sec filter is too long? I wanted to filter out the triggering of the sensor due to mechanical glitches, vibrations, etc...which happened several times in the past..

**Comment by tiaanv** [comment]
I fully agree with your strategy, especially if klipper is not catering for those false triggers.  I think disabling the built-in feature should really be an advanced function, for the user that knows WHY they are disabling it.  Would I be able to program the onboard IC, with a normal fdti serial type interface, or do I require specialised hardware? Are the uart/programming pins exposed?  I really have not looked at the schematic or board ;)

**Comment by RobertLorincz** [comment]
The micro is an ATiny412 and has UDPI programming interface. In principle, I can implement that change you are mentioning, but you will need an UDPI-capable programming device. If you have something like that, we can do a try and see how we can make it work.

**Comment by tiaanv** [comment]
Hey Robert.  That would be fantastic.  

I believe I can make up something without too much effort.  I do lots of ESP and Arduino programming.  I Can use one of my FDTI adapters with a 4.7K resistor, I think.   I'm guessing I can use the Arduino or PlatformIO IDE to do the programming.  I'm not sure if you would be willing to share the code, and I can even do the coding and share it back with you.  I am a programmer by trade.  But that part is up to you.

Let me know

**Comment by RobertLorincz** [comment]
I have no clue if FDI can be used; probably yes...I use an ATMEL ICE programming device... but in the end, it does not matter what you use; you just should know the protocol. I can provide the source code, but I'm afraid it will not be so useful. It is written in Code VIsion AVR with license. not so easy to move that code to PlatformIO or Arduino environment... at least I'm not so good in programming :D. If you provide me an email I can provide you the source code. But probably it is easiest to make the changes on my side and you try to upload the firmware on your side. I can tell you how to connect the programmer and how to set the micro into programming mode.

**Comment by tiaanv** [comment]
Agreed. not going to get code vision AVR.   If you would perhaps make the changes, and I can try and upload it, that would be great.

**Comment by tiaanv** [comment]
perfect.  tiaan@qwix.net.   I will be removing my email address, once you reply

**Comment by tiaanv** [comment]
Hi Robert.  Any joy on this?  I also bought a UDPI programmer.

**Comment by piotrb** [comment]
Hey folks, I just encountered this same problem with using this sensor with HappyHare, HH uses this sensor for a bunch of use cases which require fairly immediate triggering, its used in calibration scenarios to measure bowden lengths, to measure hotend filament path dimensions, and generally to know things like, did I manage to yank out the filament from the hotend when swapping spools, so its pretty important to get the timing right here .. in a second a LOT of filament can move by :) 

Has there been any updates on this, I'm quite annoyed I'm being forced to look for a replacement for this sensor, I very much like its base form factor :)

**Comment by tiaanv** [comment]
Unfortunately. No response from Robert.  I could decode everything there is to know about the IC and cirrcuit, and in the end wrote my own code.  Unfortunately I could not upload it using my UDPI programmer I had.  I have since bought another one just to test as well as some replacement ICs. I have not tried yet.  I will try tomorrow and let you know.  The standard programmer did not work, so this is the HV version.  If in circuit programming does not work because of some resistors in the circuit, I will try programming one outside of the circuit, but that won't be a likely solution you can follow.

**Comment by piotrb** [comment]
I literally just took some time today to use the general design of things from the CAD and try my hand at a remix that kind of keeps the general shape, but turns it into a basic D2F switch based entry sensor .. its not the easiest print .. so I may have to take a few more stabs at it .. but if I like how it turns out I'll publish it as a simpler option .. heh

I originally hoped to re-use the housing, but there is very little space in there .. so unfortunately it won't just let me shove the switch inside that :)

**Comment by moggieuk** [comment]
Just got referred to this thread...  Happy Hare uses the toolhead sensor ( the one past the extruder entry) for automatic toolhead calibration in addition to smart loading and purging.  The extruder (entry) sensor is also used but less so for calibration other than to measure distance between toolhead and extruder sensors.

A delay in sensor operation would be an issue since it obfuscates the exact measurement.

Since I have not experience with this extruder yet I can't offer more advice, but Happy Hare does have all the "intelligence" already.  I don't think a double trigger would be a practical problem with HH since it timestamps all sensor changes and ignores those that don't make sense.

Ideally if there were a way to turn off the "intelligent" features and get immediate feedback it would likely be 100% compatible..

**Comment by tiaanv** [comment]
Thx for the input @moggieuk 

The reason I started this thread and engaged with @RobertLorincz was exactly because of the info you supplied above.  
I don't really think the design goal of Robert was for this to be used in a MMU setup, but I had this before deciding to go the MMU route, and subsequently insisted on making it useful ;)

For anyone else reading:

I quickly figured out that the very long debounce/filter delay would render it useless in a timing-sensitive application such as Happy Hare.  I wrote alternative firmware for the sensor, but in-place programming is impossible from my tests.  I have tried two different programmers (HV and non-HV).  In the end, whilst trying to program it I managed to toast the chip. I have a set of replacement ICs, but I got lazy, and simply hacked the PCB to make it "dumb".  I don't necessarily recommend this, as it is quite finicky, and you have to bridge and remove a few tracks/components.   I sacrificed the RGB LEDs, but have two useful sensors: Extruder entry, and I also use the tangle (tension switch) sensor for extruder sync.  I think that's a win.  Ultimately this is not a replacement for the tool head sensor, and I strongly recommend that multi-material users have one for reliability with MMU/HH

I am positive that @RobertLorincz will update his firmware for the next iteration of the Orbiter Sensor to be able to disable the software debounce-filtering (delay), until then simply ignore the sensors of the orbiter, and opt for a toolhead sensor in whatever toolhead/cutter config you have.

**Comment by RobertLorincz** [comment]
Hi, sorry for the late answer. I had some troubles in the beginning of the year and could not concentrate too much on these topics... I would recommend that for such discussion, contact me by email or via Facebook Messenger, which I follow more often than GitHub.
So there is a hardware delay implemented in the firmware. I also made a new code to reduce this delay to 0.1 seconds; probably the best implementation would be to have the filter selectable by user. The biggest issue though is the reprogramming of the devices. I will upload in the next few minutes the new modified firmare...and share the reprogramming method.

**Comment by RobertLorincz** [comment]
I uploaded the modified firmware and the description here:
https://github.com/RobertLorincz/Orbiter-2-Smart-Sensor/tree/88defb5046f2ba5a8e258898a16512df0ffe6393/Firmware

**Comment by markleaf131313** [comment]
The lowest price I could find the programmer you mentioned was for ~$150. Idk if you want to add but you can create one with an Arudino Nano (with mega328p), a resistor, and capacitor for ~$5. https://github.com/ElTangas/jtag2updi

I also have the Mellow 2.2 version, which used a different chip Tiny202, and needed the tangle switch and unload button pressed for 4 seconds to enter the dev mode.

**Comment by piotrb** [comment]
@markleaf131313 did you try the new firmware with the ATiny202 or is that a bust?

**Comment by markleaf131313** [comment]
@piotrb Yes it flashed, just needed to choose the Tiny202. I haven't hooked it up yet, but testing the filament switch after the flash it no longer had as long of a delay.

**Comment by piotrb** [comment]
How's your ability to use it with happyhare? While it does seem better not sure if that was good enough .. 0.1s is still a bit of filament movement

**Comment by markleaf131313** [comment]
I can't test as my thermistor stopped working and I'm waiting for a replacement. It would depend what you use it for, if you are using it to home to and you are going fast at something like 200mm/s, then that's still 20mm of filament that would move past it. Not sure if there's that much room between it and the gears in the extruder.

@RobertLorincz Would it be possible to disable the filtering all together in an extra firmware? It doesn't have to be the default firmware sold on the hardware, you can create the UDPI programmer pretty easily, it would just be for people like us that need the 0 ms delay.

**Comment by alexinnnn** [comment]
Hi everyone,
I encountered a similar problem with a large delay in this sensor when working with happyhare. I built a programmer based on arduino nano with the jtag2updi project. But now I can't switch the sensor into programming mode - I press the button for 4 seconds and nothing happens( I tried pressing the button and holding when I turn on the sensor, but also the same situation, it is just red light.
When the sensor is connected to the computer and I try to execute a simple command "avrdude-c jtag2updi -P com4 -p t202" (I have ATtiny202). It shows:
Warning: bad response to enter progmode command: RSP_ILLEGAL_MCU_STATE
Warning: bad response to enter progmode command: RSP_ILLEGAL_MCU_STATE
Error: bad response to read memory command: RSP_ILLEGAL_MCU_STATE
Error: bad response to read memory command: RSP_ILLEGAL_MCU_STATE
Error: unable to read byte at address 0x0000
Error: unable to read signature data for part ATtiny202 (rc = -2)
Error: unable to read signature data (rc = -2)
use -F to override this check
Error: bad response to leave progmode command: RSP_ILLEGAL_MCU_STATE
Avrdude done.  Thank you.
What am I doing wrong? Help!

**Comment by alexinnnn** [comment]
A few hours later...
I noticed that it "needed the tangle switch and unload button pressed for 4 seconds to enter the dev mode". Nice.
So the final command "avrdude -p t202 -c jtag2updi -P com4 -U flash:w:fw.hex" have successfully flashed it. No need to set fuses.
I will test it, but I think it should have an option for 0ms delay. It may have been better.

**Comment by RobertLorincz** [comment]
I hope the fuses remained the same as before. Good to hear you managed to reflash it. The current filter time is 100ms. I can still decrease it, but I would avoid that because these are mechanical switches; over time they generate noise when switching. So a debounce filter is a must. Even the filament insertion can trigger it twice, so this small filter will improve on that. Let me know how it works.
Yeap, you need to enter in programming mode. As I described earlier, this is because I share the UDPI signal with an outpout signal, so they would normally conflict. The entering into programming mode esentially puts the sensor output into HZ mode and allows the pin to be controlled by the programmer.

**Comment by robotfishe** [comment]
Question @RobertLorincz - are current versions of the sensor shipping with the updated firmware? I purchased mine in late November 2025.
